actix-session/src/storage/redis_actor.rs

```diff
@@ -0,0 +1,294 @@
+            .addr
+            .send(cmd)
+            .await
+            .map_err(Into::into)
+            .map_err(SaveError::Other)?
+            .map_err(Into::into)
+            .map_err(SaveError::Other)?;
+
+        match result {
+            RespValue::SimpleString(_) => Ok(session_key),
+            RespValue::Nil => Err(SaveError::Other(anyhow::anyhow!(
+                "Failed to save session state. A record with the same key already existed in Redis"
+            ))),
+            err => Err(SaveError::Other(anyhow::anyhow!(
+                "Failed to save session state. {:?}",
+                err
+            ))),
+        }
+    }
+
+    async fn update(
+        &self,
+        session_key: SessionKey,
+        session_state: SessionState,
+        ttl: &Duration,
+    ) -> Result<SessionKey, UpdateError> {
+        let body = serde_json::to_string(&session_state)
+            .map_err(Into::into)
+            .map_err(UpdateError::Serialization)?;
+        let cache_key = (self.configuration.cache_keygen)(session_key.as_ref());
+
+        let cmd = Command(resp_array![
+            "SET",
+            cache_key,
+            body,
+            "XX", // XX: Only set the key if it already exist.
+            "EX", // EX: set expiry
+            format!("{}", ttl.whole_seconds())
+        ]);
+
+        let result = self
+            .addr
+            .send(cmd)
+            .await
+            .map_err(Into::into)
+            .map_err(UpdateError::Other)?
+            .map_err(Into::into)
+            .map_err(UpdateError::Other)?;
+
+        match result {
+            RespValue::Nil => {
+                // The SET operation was not performed because the XX condition was not verified.
+                // This can happen if the session state expired between the load operation and the
+                // update operation. Unlucky, to say the least. We fall back to the `save` routine
+                // to ensure that the new key is unique.
+                self.save(session_state, ttl)
+                    .await
+                    .map_err(|err| match err {
+                        SaveError::Serialization(err) => UpdateError::Serialization(err),
+                        SaveError::Other(err) => UpdateError::Other(err),
+                    })
+            }
+            RespValue::SimpleString(_) => Ok(session_key),
+            val => Err(UpdateError::Other(anyhow::anyhow!(
+                "Failed to update session state. {:?}",
+                val
+            ))),
+        }
+    }
+
+    async fn delete(&self, session_key: &SessionKey) -> Result<(), anyhow::Error> {
+        let cache_key = (self.configuration.cache_keygen)(session_key.as_ref());
+
+        let res = self
+            .addr
+            .send(Command(resp_array!["DEL", cache_key]))
+            .await?;
+
+        match res {
+            // Redis returns the number of deleted records
+            Ok(RespValue::Integer(x)) if x > 0 => Ok(()),
```

Comment:
Doing this check is a difference in behavior between this and redis_rs impl. I can't think of a codepath that would get here from a high level; I think you'd have to forge a session. But, for example, this test fails:

```rust
    #[actix_web::test]
    async fn remove_non_existent_session() {
        let store = redis_actor_store();
        let session_key = generate_session_key();
        store.delete(&session_key).await.unwrap();
    }
```

Reply:
The only case where this might happen is if the session expires in Redis while a request is being handled. I think we can remove it.

Reply:
Yes please. I think they should be the same behaviorally from the session perspective.

Reply:
Done in [fd23949](https://github.com/actix/actix-extras/pull/212/commits/fd2394933f38edb06b885c0d2a0b8fbc7d34d52e)